### PR TITLE
P4 1365 basic auth on webhooks

### PR DESCRIPTION
**Pull request by martyn-w**
### Jira link

[P4-1365](https://dsdmoj.atlassian.net/browse/P4-1365)

### What?

I have added/removed/altered:

- Added support for basic auth on webhooks

### Why?

I am doing this because:

- suppliers have requested it

### Have you? (optional)

- [X] Updated webhook docs

### Deployment risks (optional)

- Includes destructive/migratory actions which may effect critical offender data
- Changes an api that is used in production

### CODE REVIEW

File: app/jobs/notify_webhook_job.rb [hunk 1]
```diff
@@ -14,9 +14,13 @@ def perform(notification_id:)
     begin
       data = ActiveModelSerializers::Adapter.create(NotificationSerializer.new(notification)).to_json
       hmac = Encryptor.hmac(notification.subscription.secret, data)
```
Comment (smoothcontract) [inline app/jobs/notify_webhook_job.rb:16]: Maybe nice to assign `subscription = notification.subscription` here (or actually the line above) to simplify access throughout the method

File: app/jobs/notify_webhook_job.rb [hunk 2]
```diff
@@ -29,10 +33,4 @@ def perform(notification_id:)
       raise e # re-raise the error to force the notification to be retried by sidekiq later
     end
   end
-
-private
-
-  def client
```
Comment (smoothcontract) [inline app/jobs/notify_webhook_job.rb:35]: I think I prefer this as a separate method TBH - makes the main part of the code easier to read.

Comment (smoothcontract) [reply app/jobs/notify_webhook_job.rb:35]: You could perhaps pass `subscription` as an argument to this method and move the basic auth header assignment stuff (lines 18-20) to here:

```ruby
def client(subscription)
  Faraday.new(headers: { 'Content-Type': 'application/vnd.api+json', 'User-Agent': 'pecs-webhooks/v1' }).tap do |client|
    if subscription.username.present? && subscription.password.present?
      client.headers['Authorization'] = "Basic #{Base64.strict_encode64(subscription.username + ':' + subscription.password)}"
    end
  end
end
```

File: app/services/encryptor.rb [hunk 3]
```diff
@@ -20,7 +20,7 @@ def self.encrypt(value)
   end
 
   def self.decrypt(value)
-    new.decrypt_and_verify(value)
+    new.decrypt_and_verify(value) if value.present?
```
Comment (smoothcontract) [inline app/services/encryptor.rb:23]: Does this change have coverage?

File: app/jobs/notify_webhook_job.rb (hunk 1 above)
Comment (martyn-w) [reply app/jobs/notify_webhook_job.rb]: 👍 yep that cleans it up a little, good suggestion

File: app/services/encryptor.rb (hunk 3 above)
Comment (martyn-w) [reply app/services/encryptor.rb:23]: it does now :-)

File: app/jobs/notify_webhook_job.rb (hunk 2 above)
Comment (martyn-w) [reply app/jobs/notify_webhook_job.rb:35]: Ok - i've done this minus the `tap`

**Pull request merged**